distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,782 @@
+        self._frame_nbytes_remaining = 0
+
+    def connection_lost(self, exc=None):
+        if exc is None:
+            exc = CommClosedError("Connection closed")
+        self._exception = exc
+        self._transport = None
+        self._is_closed.set_result(None)
+
+        # Unblock read, if any
+        self._queue.put_nowait(_COMM_CLOSED)
+
+        # Unblock write, if any
+        if self._paused:
+            waiter = self._drain_waiter
+            if waiter is not None:
+                self._drain_waiter = None
+                if not waiter.done():
+                    waiter.set_exception(exc)
+
+    def pause_writing(self):
+        self._paused = True
+
+    def resume_writing(self):
+        self._paused = False
+
+        waiter = self._drain_waiter
+        if waiter is not None:
+            self._drain_waiter = None
+            if not waiter.done():
+                waiter.set_result(None)
+
+    async def read(self):
+        # Even if comm is closed, we still yield all received data before
+        # erroring
+        if self._queue is not None:
+            out = await self._queue.get()
+            if out is not _COMM_CLOSED:
+                return out
+            self._queue = None
+        raise self._exception
+
+    async def write(self, frames):
+        if self._exception:
+            raise self._exception
+
+        nframes = len(frames)
+        frames_nbytes = [nbytes(f) for f in frames]
+        # TODO: the old TCP comm included an extra `msg_nbytes` prefix that
+        # isn't really needed. We include it here for backwards compatibility,
+        # but this could be removed if we ever want to make another breaking
+        # change to the comms.
+        msg_nbytes = sum(frames_nbytes) + (nframes + 1) * 8
+        header = struct.pack(f"{nframes + 2}Q", msg_nbytes, nframes, *frames_nbytes)
+        frames = [header, *frames]
+
+        if msg_nbytes < 2 ** 17:  # 128kiB
+            # small enough, send in one go
+            frames = [b"".join(frames)]
+
+        if len(frames) > 1:
+            self._transport.writelines(frames)
+        else:
+            self._transport.write(frames[0])
+        if self._transport.is_closing():
+            await asyncio.sleep(0)
+        elif self._paused:
+            self._drain_waiter = self._loop.create_future()
+            await self._drain_waiter
+
+        return msg_nbytes
+
+
+class TCP(Comm):
+    max_shard_size = dask.utils.parse_bytes(dask.config.get("distributed.comm.shard"))
+
+    def __init__(
+        self,
+        protocol,
+        local_addr: str,
+        peer_addr: str,
+        deserialize: bool = True,
+    ):
+        self._protocol = protocol
+        self._local_addr = local_addr
+        self._peer_addr = peer_addr
+        self.deserialize = deserialize
+        self._closed = False
+        super().__init__()
+
+        # setup a finalizer to close the protocol if the comm was never explicitly closed
```

Comment:
Why not just a `__del__` method?

Reply:
That would work too. I've borrowed this from the other comm implementations. I suspect we could move all to `__del__` methods - in python 2 there was a higher chance a `__del__` would never be called due to a cycle, but this is no longer the case in python 3. Leaving as is for now though.